pkg/status/workstatus.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const workStatusIdentificationIndexKey = "workStatusIdentificationIndex"
+
+type workStatus struct {
+	workStatusRef
+	status map[string]interface{}
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, ref workStatusRef) error {
+	workStatus := &workStatus{
+		workStatusRef: ref, //readonly
+		status:        nil,
+	}
+
+	obj, err := c.workStatusLister.ByNamespace(ref.wecName).Get(ref.name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get workstatus (%v): %w", ref, err)
+		}
+	} else {
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return fmt.Errorf("failed to get status from workstatus (%v): %w", ref, err)
+		}
+
+		workStatus.status = status // might be nil
+	}
+
+	combinedStatusSet := c.combinedStatusResolver.NoteWorkStatus(workStatus) // nil .status is equivalent to deleted
+	for combinedStatus := range combinedStatusSet {
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
+		return nil
+	}
+
+	// remove the status if singleton status label value is unset
+	if statusLabelVal == util.BindingPolicyLabelSingletonStatusValueUnset {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &ref.sourceObjectIdentifier, emptyStatus,
+			c.listers, c.wdsDynClient)
+	}
+
+	status, err := util.GetWorkStatusStatus(obj)
+	if err != nil {
+		return err
+	}
+
+	if status == nil {
+		// when status is updated, a new event will trigger update
+		return nil
+	}
+
+	c.logger.Info("updating singleton status", "objectIdentifier", ref.sourceObjectIdentifier)
```

Comment:
Use the logger from the context in preference to `c.logger`. That is a general design rule.